pkg/netlink/netlink_watcher.go

```diff
@@ -0,0 +1,2088 @@
+}
+
+func deepEqualN(n1 *NexthopStruct, n2 *NexthopStruct, nc bool) bool {
+	if !reflect.DeepEqual(n1.Vrf.Name, n2.Vrf.Name) {
+		return false
+	}
+	if !reflect.DeepEqual(n1.Weight, n2.Weight) {
+		return false
+	}
+	if !reflect.DeepEqual(n1.ID, n2.ID) {
+		return false
+	}
+	if !reflect.DeepEqual(n1.nexthop, n2.nexthop) {
+		return false
+	}
+	if !reflect.DeepEqual(n1.Key, n2.Key) {
+		return false
+	}
+	if !reflect.DeepEqual(n1.Local, n2.Local) {
+		return false
+	}
+	if !reflect.DeepEqual(n1.Metadata, n2.Metadata) {
+		return false
+	}
+	if !reflect.DeepEqual(n1.Metric, n2.Metric) {
+		return false
+	}
+	if !reflect.DeepEqual(n1.Scope, n2.Scope) {
+		return false
+	}
+	if !reflect.DeepEqual(n1.Resolved, n2.Resolved) {
+		return false
+	}
+	if !reflect.DeepEqual(n1.Protocol, n2.Protocol) {
+		return false
+	}
+	if !reflect.DeepEqual(n1.NhType, n2.NhType) {
+		return false
+	}
+	if nc {
+		if len(n1.RouteRefs) != len(n2.RouteRefs) {
+			return false
+		}
+		for i := range n1.RouteRefs {
+			ret := deepEqualR(n1.RouteRefs[i], n2.RouteRefs[i], false)
+			if !ret {
+				return false
+			}
+		}
+	}
+	/*if !reflect.DeepEqual(N1.Neighbor, N2.Neighbor) {
+		return false
+	}*/
+	return true
+}
+
+func deepEqualR(r1 *RouteStruct, r2 *RouteStruct, nc bool) bool {
+	if !reflect.DeepEqual(r1.Vrf.Name, r2.Vrf.Name) {
+		return false
+	}
+	if !reflect.DeepEqual(r1.Route0, r2.Route0) {
+		return false
+	}
+	if !reflect.DeepEqual(r1.Key, r2.Key) {
+		return false
+	}
+	if !reflect.DeepEqual(r1.NlType, r2.NlType) {
+		return false
+	}
+	if !reflect.DeepEqual(r1.Metadata, r2.Metadata) {
+		return false
+	}
+	if nc {
+		if len(r1.Nexthops) != len(r2.Nexthops) {
+			return false
+		}
+		return deepEqualN(r1.Nexthops[0], r2.Nexthops[0], false)
+	}
+	return true
+}
+
+func deepCheck(v1 interface{}, v2 interface{}, event []string) bool {
+	if strings.HasPrefix(event[1], "route") {
+		r1 := v1.(*RouteStruct)
+		r2 := v2.(*RouteStruct)
+		return deepEqualR(r1, r2, true)
+	} else if strings.HasPrefix(event[1], "nexthop") {
+		n1 := v1.(*NexthopStruct)
+		n2 := v2.(*NexthopStruct)
+		return deepEqualN(n1, n2, true)
+	}
+	return true
+}
+
+// nolint
+func notify_changes(new_db map[interface{}]interface{}, old_db map[interface{}]interface{}, event []string) {
+	DB2 := old_db
+	DB1 := new_db
+	/* Checking the Updated entries in the netlink db by comparing the individual keys and their corresponding values in old and new db copies
+	   entries with same keys with different values and send the notification to vendor specific module */
+	for k1, v1 := range DB1 {
+		for k2, v2 := range DB2 {
+			if k1 == k2 {
+				if !deepCheck(v1, v2, event) {
+					// To Avoid in-correct update notification due to race condition in which metadata is nil in new entry and crashing in dcgw module
+					if strings.Contains(event[1], "route") || strings.HasPrefix(event[1], "nexthop") {
+						var rv *RouteStruct
+						var nv *NexthopStruct
+						if strings.Contains(event[1], "route") {
```

Comment:
how do you know that you have 2 elements in event array?

Reply:
done

Reply:
not done. I see
```Go
if rv.Vrf.Status.VrfOperStatus == infradb.VrfOperStatusToBeDeleted {
   notifyAddDel(rv, event.Operation[2])
````

What does index 2 mean here? Why was it 1? Would it make sense to create a dedicated struct where we ahve 3 members and their name clearly explain what those events mean?

Reply:
each event will have 3 operations add, update and delete, accordingly we are filling operations for each event type, so that's where we have length 3 and index 2 for delete, which already defined as constants

Reply:
For me, it is just a magic number. You might have context how indexing in this array correspond to operations, but I do not have such a context and people who come and try to understand this code, does not have this context either.
Why we cannot use?
```Go
notifyAddDel(rv, event.operations.to_delete)
```

Plus

The code chunk at the moment look like
```Go
if event.EventType == ROUTE || event.EventType == NEXTHOP {
var rv *RouteStruct
var nv *NexthopStruct
if event.EventType == ROUTE {
	rv, ok = v1.(*RouteStruct)
	if !ok {
		log.Printf("Netlink: Type mismatch")
		continue
	}
	if rv.Vrf.Status.VrfOperStatus == infradb.VrfOperStatusToBeDeleted {
		notifyAddDel(rv, event.Operation[2])
		delete(new_db, k1)
		delete(old_db, k1)
		continue
	}
} else if event.EventType == NEXTHOP {
	nv, ok = v1.(*NexthopStruct)
	if !ok {
		log.Printf("Netlink: Type mismatch")
		continue
	}
	if nv.Vrf.Status.VrfOperStatus == infradb.VrfOperStatusToBeDeleted {
		notifyAddDel(nv, event.Operation[2])
		delete(new_db, k1)
		delete(old_db, k1)
		continue
	}
}
```

Don't you have a feeling that the code in first if is very similar in the code in the second if?
Can we have an interface which would return the status, type assert to that interface once, get the status and do your actions without these duplications? or any other means

Reply:
Done

Reply:
This
```Go
switch event.EventType {
case ROUTE:
	status, ok = v1.(VrfStatusGetter)
case NEXTHOP:
	status, ok = v1.(VrfStatusGetter)
default:
}
```
to
```Go
status, ok := v1.(VrfStatusGetter)
```

Reply:
Done